libs/httpsignature/httpsignature.go

```diff
@@ -107,19 +114,47 @@ func (sp *SignatureParams) IsMalformed() bool {
 
 // BuildSigningString builds the signing string according to the SignatureParams s and
 // HTTP request req
-// TODO Add support for digest generation based on req.Body?
 func (sp *SignatureParams) BuildSigningString(req *http.Request) (out []byte, err error) {
+	if req.Body != nil {
+		body, err := requestutils.Read(context.Background(), req.Body)
+		if err != nil {
+			return nil, err
+		}
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		return sp.buildSigningString(body, req.Header, req)
+	}
+	return sp.buildSigningString(nil, req.Header, req)
+}
+
+// BuildSigningStringForResponse builds the signing string according to the SignatureParams s and
+// HTTP response resp
+func (sp *SignatureParams) BuildSigningStringForResponse(resp *http.Response) (out []byte, err error) {
+	if resp.Body != nil {
+		body, err := requestutils.Read(context.Background(), resp.Body)
+		if err != nil {
+			return out, err
+		}
+		resp.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		return sp.buildSigningString(body, resp.Header, resp.Request)
+	}
+	return sp.buildSigningString(nil, resp.Header, resp.Request)
+}
+
+func (sp *SignatureParams) buildSigningString(body []byte, headers http.Header, req *http.Request) (out []byte, err error) {
 	if sp.IsMalformed() {
 		return nil, errors.New("refusing to build signing string with malformed params")
 	}
 
-	headers := sp.Headers
-	if len(headers) == 0 {
-		headers = []string{"date"}
+	signedHeaders := sp.Headers
+	if len(signedHeaders) == 0 {
+		signedHeaders = []string{"date"}
 	}
 
-	for i, header := range headers {
+	for i, header := range signedHeaders {
 		if header == RequestTargetHeader {
+			if req == nil {
```

Comment:
Could this be moved the top of the function so we fail early? Then the other nil check [here](https://github.com/brave-intl/bat-go/pull/2000/files#diff-d172bed3fee7b085798087f7a1ac3e31c22d614aab41ac1a75a5b7730dbead25R179) can be removed